// ngx-basic-primitives/esm2020/lib/diagrams/components/checkbox.component.mjs
import { Component, Input } from "@angular/core";
import * as i0 from "@angular/core";
export class CheckboxComponent {
    constructor() {
        this.id = null;
        this.label = "Selected";
        this.isSelected = false;
        this.onChange = () => null;
        this.borderWidth = "1px";
    }
}
CheckboxComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: CheckboxComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
CheckboxComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.2", type: CheckboxComponent, selector: "bp-checkbox", inputs: { id: "id", label: "label", isSelected: "isSelected" }, ngImport: i0, template: `<div class="checkbox-div">
    <input type="checkbox" name="checkbox" class="checkbox-input" [checked]="isSelected" (change)="onChange" [attr.data-id]="id" />
      <span name="selectiontext" class="checkbox-span" [attr.data-id]="id" >
        {{label}}
      </span>
    </div>`, isInline: true, styles: [".checkbox-div{text-align:left}.checkbox-input{font-size:12px;line-height:12px;vertical-align:middle;margin:0;padding:0}.checkbox-span{font-size:12px;line-height:12px;vertical-align:middle;margin:0;padding:0 0 0 2px}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: CheckboxComponent, decorators: [{
            type: Component,
            args: [{ selector: 'bp-checkbox', template: `<div class="checkbox-div">
    <input type="checkbox" name="checkbox" class="checkbox-input" [checked]="isSelected" (change)="onChange" [attr.data-id]="id" />
      <span name="selectiontext" class="checkbox-span" [attr.data-id]="id" >
        {{label}}
      </span>
    </div>`, styles: [".checkbox-div{text-align:left}.checkbox-input{font-size:12px;line-height:12px;vertical-align:middle;margin:0;padding:0}.checkbox-span{font-size:12px;line-height:12px;vertical-align:middle;margin:0;padding:0 0 0 2px}\n"] }]
        }], propDecorators: { id: [{
                type: Input
            }], label: [{
                type: Input
            }], isSelected: [{
                type: Input
            }] } });
;
