// ngx-basic-primitives/esm2020/lib/diagrams/components/cursor.component.mjs
import { Component, Input } from "@angular/core";
import { TemplateConfig } from "../../configs/template-config";
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class CursorComponent {
    constructor() {
        this.templateConfig = new TemplateConfig();
        this.borderWidth = "1px";
        this.width = "0px";
        this.height = "0px";
    }
    ngOnChanges(changes) {
        const { cursorPadding, itemSize, cursorBorderWidth } = this.templateConfig;
        this.width = (itemSize.width + cursorPadding.left + cursorPadding.right) + "px",
            this.height = (itemSize.height + cursorPadding.top + cursorPadding.bottom) + "px",
            this.borderWidth = cursorBorderWidth + "px";
    }
}
CursorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: CursorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
CursorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.2", type: CursorComponent, selector: "bp-cursor", inputs: { templateConfig: "templateConfig" }, usesOnChanges: true, ngImport: i0, template: '<div class="cursor" [ngStyle]="{width, height, borderWidth}"></div>', isInline: true, styles: [".cursor{position:absolute;font-family:Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;-webkit-tap-highlight-color:rgba(0,0,0,0);-webkit-user-select:none;-webkit-touch-callout:none;user-select:none;box-sizing:content-box;border-radius:4px;border:2px solid #fbd850;color:#eb8f00;width:100%;height:100%;left:-2px;top:-2px}\n"], directives: [{ type: i1.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: CursorComponent, decorators: [{
            type: Component,
            args: [{ selector: 'bp-cursor', template: '<div class="cursor" [ngStyle]="{width, height, borderWidth}"></div>', styles: [".cursor{position:absolute;font-family:Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;-webkit-tap-highlight-color:rgba(0,0,0,0);-webkit-user-select:none;-webkit-touch-callout:none;user-select:none;box-sizing:content-box;border-radius:4px;border:2px solid #fbd850;color:#eb8f00;width:100%;height:100%;left:-2px;top:-2px}\n"] }]
        }], propDecorators: { templateConfig: [{
                type: Input
            }] } });
;
