// ngx-basic-primitives/esm2020/lib/diagrams/components/dot-highlight.component.mjs
import { Component, Input } from "@angular/core";
import { TemplateConfig } from "../../configs/template-config";
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class DotHighlightComponent {
    constructor() {
        this.templateConfig = new TemplateConfig();
        this.divStyle = {};
    }
    ngOnChanges(changes) {
        const { minimizedItemCornerRadius, highlightPadding, highlightBorderWidth, minimizedItemSize } = this.templateConfig;
        let radius = 0;
        if (minimizedItemCornerRadius === null) {
            radius = Math.max(minimizedItemSize.width / 2, minimizedItemSize.height / 2) + highlightPadding.left;
        }
        else {
            radius = minimizedItemCornerRadius + highlightPadding.left;
        }
        this.divStyle = {
            borderWidth: highlightBorderWidth + "px",
            left: -highlightBorderWidth + "px",
            top: -highlightBorderWidth + "px",
            "-moz-border-radius": radius + "px",
            "-webkit-border-radius": radius + "px",
            "-khtml-border-radius": radius + "px",
            "border-radius": radius + "px"
        };
    }
}
DotHighlightComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: DotHighlightComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
DotHighlightComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.2", type: DotHighlightComponent, selector: "bp-dot-highlight", inputs: { templateConfig: "templateConfig" }, usesOnChanges: true, ngImport: i0, template: '<div class="dot-highlight" [ngStyle]="divStyle"></div>', isInline: true, styles: [".dot-highlight{position:absolute;font-family:Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;-webkit-tap-highlight-color:rgba(0,0,0,0);-webkit-user-select:none;-webkit-touch-callout:none;user-select:none;box-sizing:content-box;border:1px solid #fbcb09;background:#fdf5ce;color:#c77405;width:100%;height:100%;left:0px;top:0px}\n"], directives: [{ type: i1.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: DotHighlightComponent, decorators: [{
            type: Component,
            args: [{ selector: 'bp-dot-highlight', template: '<div class="dot-highlight" [ngStyle]="divStyle"></div>', styles: [".dot-highlight{position:absolute;font-family:Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;-webkit-tap-highlight-color:rgba(0,0,0,0);-webkit-user-select:none;-webkit-touch-callout:none;user-select:none;box-sizing:content-box;border:1px solid #fbcb09;background:#fdf5ce;color:#c77405;width:100%;height:100%;left:0px;top:0px}\n"] }]
        }], propDecorators: { templateConfig: [{
                type: Input
            }] } });
;
