// ngx-basic-primitives/esm2020/lib/diagrams/components/group-title.component.mjs
import { Component, Input } from "@angular/core";
import { OrgItemConfig } from "../../configs/org-item-config";
// @ts-ignore
import { highestContrast } from "basicprimitives";
import { HorizontalAlignmentType, TextOrientationType, VerticalAlignmentType } from "../../enums";
import * as i0 from "@angular/core";
import * as i1 from "../rotated-text.component";
import * as i2 from "@angular/common";
class GroupTitleOptions {
    constructor(config = {}) {
        this.groupTitleFontSize = "12px";
        this.groupTitleFontFamily = "Arial";
        this.groupTitleFontWeight = "normal";
        this.groupTitleFontStyle = "normal";
        this.itemTitleSecondFontColor = "#000080";
        this.itemTitleFirstFontColor = "#ffffff";
        this.groupTitleOrientation = TextOrientationType.RotateRight;
        this.groupTitleHorizontalAlignment = HorizontalAlignmentType.Center;
        this.groupTitleVerticalAlignment = VerticalAlignmentType.Middle;
        this.groupTitleColor = "#4169e1";
        Object.assign(this, config);
    }
}
export class GroupTitleComponent {
    constructor() {
        this.options = new GroupTitleOptions();
        this.itemConfig = new OrgItemConfig();
        this.width = 0;
        this.height = 0;
        this.divStyle = null;
        this.label = "";
    }
    ngOnChanges(changes) {
        const backgroundColor = this.itemConfig.groupTitleColor || this.options.groupTitleColor;
        const color = highestContrast(backgroundColor, this.options.itemTitleSecondFontColor, this.options.itemTitleFirstFontColor);
        this.divStyle = {
            backgroundColor,
            color,
            fontSize: this.options.groupTitleFontSize,
            fontFamily: this.options.groupTitleFontFamily,
            fontWeight: this.options.groupTitleFontWeight,
            fontStyle: this.options.groupTitleFontStyle
        };
        this.label = (this.itemConfig.groupTitle || "").replace("\n", "<br/>");
    }
}
GroupTitleComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: GroupTitleComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
GroupTitleComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.2", type: GroupTitleComponent, selector: "bp-group-title", inputs: { options: "options", itemConfig: "itemConfig", width: "width", height: "height" }, usesOnChanges: true, ngImport: i0, template: "<div class=\"group-title\" [ngStyle]=\"divStyle\">\r\n  <rotated-text\r\n    [width]=\"width\"\r\n    [height]=\"height\"\r\n    [orientation]=\"options.groupTitleOrientation\"\r\n    [horizontalAlignment]=\"options.groupTitleHorizontalAlignment\"\r\n    [verticalAlignment]=\"options.groupTitleVerticalAlignment\"\r\n  >{{label}}</rotated-text>\r\n</div>\r\n", styles: [".group-title{position:absolute;font-family:Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;-webkit-tap-highlight-color:rgba(0,0,0,0);-webkit-user-select:none;-webkit-touch-callout:none;user-select:none;box-sizing:content-box;border-radius:4px;background:royalblue;border-width:0;color:#fff;padding:0;width:100%;height:100%;left:-1px;top:-1px}\n"], components: [{ type: i1.RotatedText, selector: "rotated-text", inputs: ["width", "height", "orientation", "horizontalAlignment", "verticalAlignment"] }], directives: [{ type: i2.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: GroupTitleComponent, decorators: [{
            type: Component,
            args: [{ selector: 'bp-group-title', template: "<div class=\"group-title\" [ngStyle]=\"divStyle\">\r\n  <rotated-text\r\n    [width]=\"width\"\r\n    [height]=\"height\"\r\n    [orientation]=\"options.groupTitleOrientation\"\r\n    [horizontalAlignment]=\"options.groupTitleHorizontalAlignment\"\r\n    [verticalAlignment]=\"options.groupTitleVerticalAlignment\"\r\n  >{{label}}</rotated-text>\r\n</div>\r\n", styles: [".group-title{position:absolute;font-family:Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;-webkit-tap-highlight-color:rgba(0,0,0,0);-webkit-user-select:none;-webkit-touch-callout:none;user-select:none;box-sizing:content-box;border-radius:4px;background:royalblue;border-width:0;color:#fff;padding:0;width:100%;height:100%;left:-1px;top:-1px}\n"] }]
        }], propDecorators: { options: [{
                type: Input
            }], itemConfig: [{
                type: Input
            }], width: [{
                type: Input
            }], height: [{
                type: Input
            }] } });
;
