// ngx-basic-primitives/esm2020/lib/diagrams/components/highlight.component.mjs
import { Component, Input } from "@angular/core";
import { TemplateConfig } from "../../configs/template-config";
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class HighlightComponent {
    constructor() {
        this.templateConfig = new TemplateConfig();
        this.borderWidth = "1px";
    }
    ngOnChanges(changes) {
        const { highlightBorderWidth } = this.templateConfig;
        this.borderWidth = highlightBorderWidth + "px";
    }
}
HighlightComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: HighlightComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
HighlightComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.2", type: HighlightComponent, selector: "bp-highlight", inputs: { templateConfig: "templateConfig" }, usesOnChanges: true, ngImport: i0, template: '<div class="highlight" [ngStyle]="{borderWidth}"></div>', isInline: true, styles: [".highlight{position:absolute;font-family:Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;-webkit-tap-highlight-color:rgba(0,0,0,0);-webkit-user-select:none;-webkit-touch-callout:none;user-select:none;box-sizing:content-box;border-radius:4px;border:1px solid #fbcb09;background:#fdf5ce;color:#c77405;width:100%;height:100%;left:-1px;top:-1px}\n"], directives: [{ type: i1.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: HighlightComponent, decorators: [{
            type: Component,
            args: [{ selector: 'bp-highlight', template: '<div class="highlight" [ngStyle]="{borderWidth}"></div>', styles: [".highlight{position:absolute;font-family:Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;-webkit-tap-highlight-color:rgba(0,0,0,0);-webkit-user-select:none;-webkit-touch-callout:none;user-select:none;box-sizing:content-box;border-radius:4px;border:1px solid #fbcb09;background:#fdf5ce;color:#c77405;width:100%;height:100%;left:-1px;top:-1px}\n"] }]
        }], propDecorators: { templateConfig: [{
                type: Input
            }] } });
;
