// ngx-basic-primitives/esm2020/lib/diagrams/components/level-background.component.mjs
import { Component, Input } from "@angular/core";
import { LevelAnnotationConfig } from "../../configs/level-annotation-config";
import { LineType } from "../../enums";
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
function getBorderStyle(lineType) {
    var result = null;
    switch (lineType) {
        case LineType.Dotted:
            result = "dotted";
            break;
        case LineType.Dashed:
            result = "dashed";
            break;
        default:
            result = "solid";
            break;
    }
    return result;
}
export class LevelBackgroundComponent {
    constructor() {
        this.annotationConfig = new LevelAnnotationConfig();
        this.divStyle = {};
    }
    ngOnChanges(changes) {
        this.divStyle = {
            opacity: this.annotationConfig.opacity,
            borderColor: this.annotationConfig.borderColor,
            backgroundColor: this.annotationConfig.fillColor,
            borderWidth: this.annotationConfig.lineWidth.toString(),
            borderStyle: getBorderStyle(this.annotationConfig.lineType),
        };
    }
}
LevelBackgroundComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: LevelBackgroundComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
LevelBackgroundComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.2", type: LevelBackgroundComponent, selector: "bp-level-background", inputs: { annotationConfig: "annotationConfig" }, usesOnChanges: true, ngImport: i0, template: '<div class="level-background" [ngStyle]="divStyle"></div>', isInline: true, styles: [".level-background{position:absolute;width:100%;height:100%}\n"], directives: [{ type: i1.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: LevelBackgroundComponent, decorators: [{
            type: Component,
            args: [{ selector: 'bp-level-background', template: '<div class="level-background" [ngStyle]="divStyle"></div>', styles: [".level-background{position:absolute;width:100%;height:100%}\n"] }]
        }], propDecorators: { annotationConfig: [{
                type: Input
            }] } });
;
