// ngx-basic-primitives/esm2020/lib/diagrams/components/level-title.component.mjs
import { Component, Input } from "@angular/core";
// @ts-ignore
import { HorizontalAlignmentType, TextOrientationType, VerticalAlignmentType } from "../../enums";
import { LevelAnnotationConfig } from "../../configs/level-annotation-config";
import * as i0 from "@angular/core";
import * as i1 from "../rotated-text.component";
import * as i2 from "@angular/common";
class LevelTitleOptions {
    constructor(config = {}) {
        this.levelTitleFontSize = "12px";
        this.levelTitleFontFamily = "Arial";
        this.levelTitleFontWeight = "normal";
        this.levelTitleFontStyle = "normal";
        this.levelTitleFontColor = "#ffffff";
        this.levelTitleOrientation = TextOrientationType.Auto;
        this.levelTitleHorizontalAlignment = HorizontalAlignmentType.Center;
        this.levelTitleVerticalAlignment = VerticalAlignmentType.Middle;
        this.levelTitleColor = "#4169e1";
        Object.assign(this, config);
    }
}
export class LevelTitleComponent {
    constructor() {
        this.options = new LevelTitleOptions();
        this.annotationConfig = new LevelAnnotationConfig();
        this.width = 0;
        this.height = 0;
        this.divStyle = null;
        this.label = "";
    }
    ngOnChanges(changes) {
        const backgroundColor = this.annotationConfig.titleColor || this.options.levelTitleColor;
        const color = this.annotationConfig.titleFontColor || this.options.levelTitleFontColor;
        this.divStyle = {
            backgroundColor,
            color,
            fontSize: this.options.levelTitleFontSize,
            fontFamily: this.options.levelTitleFontFamily,
            fontWeight: this.options.levelTitleFontWeight,
            fontStyle: this.options.levelTitleFontStyle
        };
        this.label = (this.annotationConfig.title || "").replace("\n", "<br/>");
    }
}
LevelTitleComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: LevelTitleComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
LevelTitleComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.2", type: LevelTitleComponent, selector: "bp-level-title", inputs: { options: "options", annotationConfig: "annotationConfig", width: "width", height: "height" }, usesOnChanges: true, ngImport: i0, template: "<div class=\"level-title\" [ngStyle]=\"divStyle\">\r\n  <rotated-text\r\n    [width]=\"width\"\r\n    [height]=\"height\"\r\n    [orientation]=\"options.levelTitleOrientation\"\r\n    [horizontalAlignment]=\"options.levelTitleHorizontalAlignment\"\r\n    [verticalAlignment]=\"options.levelTitleVerticalAlignment\"\r\n  >{{label}}</rotated-text>\r\n</div>", styles: [".level-title{position:absolute;font-family:Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;-webkit-tap-highlight-color:rgba(0,0,0,0);-webkit-user-select:none;-webkit-touch-callout:none;user-select:none;box-sizing:content-box;border-radius:4px;background:royalblue;border-width:0;color:#fff;padding:0;width:100%;height:100%;left:-1px;top:-1px}\n"], components: [{ type: i1.RotatedText, selector: "rotated-text", inputs: ["width", "height", "orientation", "horizontalAlignment", "verticalAlignment"] }], directives: [{ type: i2.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: LevelTitleComponent, decorators: [{
            type: Component,
            args: [{ selector: 'bp-level-title', template: "<div class=\"level-title\" [ngStyle]=\"divStyle\">\r\n  <rotated-text\r\n    [width]=\"width\"\r\n    [height]=\"height\"\r\n    [orientation]=\"options.levelTitleOrientation\"\r\n    [horizontalAlignment]=\"options.levelTitleHorizontalAlignment\"\r\n    [verticalAlignment]=\"options.levelTitleVerticalAlignment\"\r\n  >{{label}}</rotated-text>\r\n</div>", styles: [".level-title{position:absolute;font-family:Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;-webkit-tap-highlight-color:rgba(0,0,0,0);-webkit-user-select:none;-webkit-touch-callout:none;user-select:none;box-sizing:content-box;border-radius:4px;background:royalblue;border-width:0;color:#fff;padding:0;width:100%;height:100%;left:-1px;top:-1px}\n"] }]
        }], propDecorators: { options: [{
                type: Input
            }], annotationConfig: [{
                type: Input
            }], width: [{
                type: Input
            }], height: [{
                type: Input
            }] } });
;
