// ngx-basic-primitives/esm2020/lib/ngx-basic-primitives.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OrgDiagramComponent } from './diagrams/org-diagram.component';
import { FamDiagramComponent } from './diagrams/fam-diagram.component';
import { RotatedText } from './diagrams/rotated-text.component';
import { ItemComponent } from './diagrams/components/item.component';
import { HighlightComponent } from './diagrams/components/highlight.component';
import { BaseDiagramComponent } from './diagrams/base-diagram.component';
import { CursorComponent } from './diagrams/components/cursor.component';
import { CheckboxComponent } from './diagrams/components/checkbox.component';
import { GroupTitleComponent } from './diagrams/components/group-title.component';
import { DotHighlightComponent } from './diagrams/components/dot-highlight.component';
import { LevelBackgroundComponent } from './diagrams/components/level-background.component';
import { LevelTitleComponent } from './diagrams/components/level-title.component';
import * as i0 from "@angular/core";
export class BasicPrimitivesModule {
}
BasicPrimitivesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: BasicPrimitivesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BasicPrimitivesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: BasicPrimitivesModule, declarations: [BaseDiagramComponent,
        OrgDiagramComponent,
        FamDiagramComponent,
        RotatedText,
        ItemComponent,
        CursorComponent,
        HighlightComponent,
        DotHighlightComponent,
        CheckboxComponent,
        GroupTitleComponent,
        LevelBackgroundComponent,
        LevelTitleComponent], imports: [CommonModule], exports: [OrgDiagramComponent,
        FamDiagramComponent,
        RotatedText] });
BasicPrimitivesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: BasicPrimitivesModule, imports: [[
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: BasicPrimitivesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        BaseDiagramComponent,
                        OrgDiagramComponent,
                        FamDiagramComponent,
                        RotatedText,
                        ItemComponent,
                        CursorComponent,
                        HighlightComponent,
                        DotHighlightComponent,
                        CheckboxComponent,
                        GroupTitleComponent,
                        LevelBackgroundComponent,
                        LevelTitleComponent
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        OrgDiagramComponent,
                        FamDiagramComponent,
                        RotatedText
                    ]
                }]
        }] });
