// ngx-basic-primitives/esm2020/lib/configs/fam-item-config.mjs
import { Enabled, TextOrientationType, PlacementType, AdviserPlacementType } from '../enums';
export class FamItemConfig {
    constructor(config = {}) {
        this.id = null;
        this.parents = [];
        this.title = null;
        this.description = null;
        this.image = null;
        this.context = null;
        this.itemTitleColor = "#4169e1";
        this.minimizedItemShapeType = null;
        this.groupTitle = null;
        this.groupTitleColor = "#4169e1";
        this.isActive = true;
        this.hasSelectorCheckbox = Enabled.Auto;
        this.hasButtons = Enabled.Auto;
        this.templateName = null;
        this.showCallout = Enabled.Auto;
        this.calloutTemplateName = null;
        this.label = null;
        this.showLabel = Enabled.Auto;
        this.labelSize = null;
        this.labelOrientation = TextOrientationType.Auto;
        this.labelPlacement = PlacementType.Auto;
        this.primaryParent = null;
        this.relativeItem = null;
        this.position = null;
        this.placementType = AdviserPlacementType.Right;
        this.matrixId = null;
        this.addToMatrix = true;
        Object.assign(this, config);
    }
}
;
