// ngx-basic-primitives/esm2020/lib/configs/org-item-config.mjs
import { Enabled, ItemType, AdviserPlacementType, ChildrenPlacementType, TextOrientationType } from '../enums';
export class OrgItemConfig {
    constructor(config = {}) {
        this.id = null;
        this.parent = null;
        this.title = null;
        this.description = null;
        this.image = null;
        this.context = null;
        this.itemTitleColor = "#4169e1";
        this.minimizedItemShapeType = null;
        this.groupTitle = null;
        this.groupTitleColor = "#4169e1";
        this.isVisible = true;
        this.isActive = true;
        this.hasSelectorCheckbox = Enabled.Auto;
        this.hasButtons = Enabled.Auto;
        this.itemType = ItemType.Regular;
        this.adviserPlacementType = AdviserPlacementType.Auto;
        this.childrenPlacementType = ChildrenPlacementType.Auto;
        this.levelOffset = null;
        this.placeAssistantsAboveChildren = Enabled.Auto;
        this.placeAdvisersAboveChildren = Enabled.Auto;
        this.templateName = null;
        this.showCallout = Enabled.Auto;
        this.calloutTemplateName = null;
        this.label = null;
        this.showLabel = Enabled.Auto;
        this.labelSize = null;
        this.labelOrientation = TextOrientationType.Auto;
        Object.assign(this, config);
    }
}
;
