// ngx-basic-primitives/esm2020/lib/configs/template-config.mjs
import { LineType } from '../enums';
import { Size, Thickness } from '../structs';
export class TemplateConfig {
    constructor(config = {}) {
        this.name = null;
        this.isActive = true;
        this.itemSize = new Size(120, 100);
        this.itemBorderWidth = 1;
        this.itemTemplate = null;
        this.minimizedItemShapeType = null;
        this.minimizedItemSize = new Size(4, 4);
        this.minimizedItemCornerRadius = null;
        this.minimizedItemLineWidth = 1;
        this.minimizedItemBorderColor = null;
        this.minimizedItemLineType = LineType.Solid;
        this.minimizedItemFillColor = null;
        this.minimizedItemOpacity = 1;
        this.highlightPadding = new Thickness(2, 2, 2, 2);
        this.highlightBorderWidth = 1;
        this.highlightTemplate = null;
        this.cursorPadding = new Thickness(3, 3, 3, 3);
        this.cursorBorderWidth = 2;
        this.cursorTemplate = null;
        this.buttonsTemplate = null;
        Object.assign(this, config);
    }
}
;
